### DISCUSSION TITLE
Is there any way to deserialize a JSON string of one class into another class?

### DISCUSSION THREAD

**Discussion by IronEaterBeast** [dotnet/runtime#101790] [Q&A]
My program has a class with many properties that store all the settings and information for a user's project, like this:

```
Class Config{
    public bool EnableYoutube { get; set; }= true;
    public bool EnableFacebook { get; set; }= true;
}
```

When the user saves, I serialize this class to JSON and save it to a text file. Conversely, when the user loads a project, I read the text file and deserialize the JSON to restore the project's settings and information.

Now I have a tricky problem: when I update the software, I may add some features and need to modify or add new properties to this class, like this:

```
Class Config{
    public bool EnableYoutube { get; set; }= true;
    
    public bool EnableTiktok { get; set; }= true;
}
```

When the user updates the software and then loads a project, the deserialization of this JSON fails because the structure of the two classes has changed.
I may use a database later, but now the project schedule is tight and I can't make too big a switch. Maybe there are better ways to save settings in a plain text file? How can I solve this problem? 

After running some tests, I figured out the problem. Since I need to use AOT compilation, I have to use source generation in System.Text.Json. The code for this part is shown below:

```
[JsonSerializable(typeof(Config))]

if (objectType == typeof(Config))
{
    option = allContext.Default.Config;
}

JsonTypeInfo? option = getOption(objectType);

var jsonString = JsonSerializer.Serialize(input!, option);
```

However, when I modify the class used for saving information, the old user save file does not match any items in the source generation, which causes an error when deserializing the JSON: "JsonTypeInfo? option equals null, which may mean that an unspecified structure cannot be JSON serialized."

This seems to mean that I need to save all versions of the project's save classes in source generation, which is a bit too much.

**Comment by elgonzo** [comment]
It is unfortunately very unclear what your problem is and what you exactly mean by saying "_the deserialization of this JSON fails_". Tell exactly what your code does (how it does the deserialization exactly), what the exact expected behavior re (de)serialization is, and the exact actual (de)serialization behavior you observed.

Note that deserialization of old config Json data (serialized from your old Config class) into an instance of your new Config class is not failing because of the serializer's default behavior. The serializer simply ignores the EnableFacebook value from the Json data and otherwise happily populates the the `Config.EnableYoutube` property with the value from the respective Json property (proof: https://dotnetfiddle.net/HyIc5T) The deserializer itself would only fail to deserialize if you you set the `JsonUnmappedMemberHandling.Disallow` option, but then of course deserialization failure is mandated and expected.

**Reply by IronEaterBeast** [reply]
"The deserialization of this JSON fails" means that all the values of previously stored properties will be lost and default properties will be used, as if there were no values from the previously saved content. Even if the "EnableYouTube" property exists in both classes, its value will not be inherited but rather a new instance of  Class Config will be obtained like: new Config();

**Reply by elgonzo** [reply]
See the dotnetfiddle link i provided in my previous comment. Note that the serializer does deserialize the EnableYouTube property.
Unless you share a complete minimal code example that demonstrates the behavior you observe, it's difficult to speculate about what your code is really doing there...

**Reply by elgonzo** [reply]
By the way, i just noticed now that you wrote "_EnableYouTube_" in your last comment, with a letter-casing that is different from the two Config classes you have shown in your report (uppercase "T" vs. lowercase "t"). If there is indeed a mismatch in letter-casing between the json property name(s) and the name(s) of the properties in your .NET classes, you can configure the serializer to do case-insensitive property name matching by setting the [`JsonSerializerOptions.PropertyNameCaseInsensitive`](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializeroptions.propertynamecaseinsensitive#system-text-json-jsonserializeroptions-propertynamecaseinsensitive) option to _true_.

**Reply by IronEaterBeast** [reply]
I am currently testing and I do not have any uppercase or lowercase issues. I now suspect that this is a problem that only occurs in AOT compilation. Because I use AOT compilation and System.Text.Json, I have defined [JsonSerializable(typeof(Config))], and I suspect that these strict restrictions may be causing the error. I'm still confused, but I am trying to figure out the reason.

**Reply by IronEaterBeast** [reply]
After running some tests, I figured out the problem. Since I need to use AOT compilation, I have to use source generation in System.Text.Json. The code for this part is shown below:

```
[JsonSerializable(typeof(Config))]

if (objectType == typeof(Config))
{
    option = allContext.Default.Config;
}

JsonTypeInfo? option = getOption(objectType);

var jsonString = JsonSerializer.Serialize(input!, option);
```

However, when I modify the class used for saving information, the old user save file does not match any items in the source generation, which causes an error when deserializing the JSON: "JsonTypeInfo? option equals null, which may mean that an unspecified structure cannot be JSON serialized." 

This seems to mean that I need to save all versions of the project's save classes in source generation, which is a bit too much.

**Reply by elgonzo** [reply]
> "JsonTypeInfo? option equals null, which may mean that an unspecified structure cannot be JSON serialized."

This smells fishy. Have you perhaps multiple `Config` types in your project in different namespaces? Could it be that you included one of those `Config` types in the STJ source generation, but when you try to obtain the appropriate JsonTypeInfo for deserialization, you are using a type object for a `Config` from a different namespace that the STJ source code gen wasn't made aware of, thus no JsonTypeInfo being available? (Or something similar along those lines...)
